### DISCUSSION TITLE
April 2022 MN GLM run results

### DISCUSSION THREAD

**Discussion by hcorson-dosch-usgs** [DOI-USGS/lake-temperature-process-models#45] [General]
Summary of latest run of 64,494 models. Can compare to [February run](https://github.com/USGS-R/lake-temperature-process-models/issues/2#issuecomment-1033243751). Timing metrics defined [in this function](https://github.com/hcorson-dosch/lake-temperature-process-models/blob/denali_glm_tests/2_run/src/run_glm3.R#L50-L196).

| p1_lake_cell_tile_xwalk_df indices      | n_lakes      | n_runs | GLM version | `retry()` max_tries | n_succeeded | n_failed | build duration (min) | runs per minute |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| 1:3583   | 3583 | 64,494        | 3.2.0a8        | 2 | 64,384        | 110 (109/110 have NA values extraction error)        | 366       | 2.45        | 

| build_time_s | attempt_time_s | retry_time _s  | glm_time _s |
| ----------- | ----------- | ----------- | ----------- |
| <table>  <tr>  <th>Subset</th> <th>Min</th>    <th>Median</th> <th>Mean</th> <th>Max</th> </tr> <tr> <td>All</td> <td>1.3</td> <td>5.2</td>     <td>5.7</td> <td>37.7</td> </tr>  <tr> <td>Succeeded</td> <td>3.1</td> <td>5.2</td>     <td>5.7</td> <td>37.7</td> </tr> <tr> <td>Failed</td> <td>1.3</td> <td>2.3</td>     <td>2.5</td> <td>4.4</td> </tr></table> | <table>  <tr>  <th>Subset</th> <th>Min</th>    <th>Median</th> <th>Mean</th> <th>Max</th> </tr> <tr> <td>All</td> <td>1.3</td> <td>5.2</td>     <td>5.7</td> <td>37.7</td> </tr>  <tr> <td>Succeeded</td> <td>3</td> <td>5.2</td>     <td>5.7</td> <td>37.7</td> </tr> <tr> <td>Failed</td> <td>1.3</td> <td>2.3</td>     <td>2.5</td> <td>4.3</td> </tr></table>        | <table>  <tr>  <th>Subset</th> <th>Min</th>    <th>Median</th> <th>Mean</th> <th>Max</th> </tr> <tr> <td>All</td> <td>1.7</td> <td>3.9</td>     <td>4.2</td> <td>31.1</td> </tr>  <tr> <td>Succeeded</td> <td>2.03</td> <td>3.9</td>     <td>4.2</td> <td>31.1</td> </tr> <tr> <td>Failed</td> <td>1.7</td> <td>2</td>     <td>2.1</td> <td>3.7</td> </tr></table>  `retry_time_s` set to `NaN` when model fails (doesn't capture time for failed result extraction)     | <table>  <tr>  <th>Subset</th> <th>Min</th>    <th>Median</th> <th>Mean</th> <th>Max</th> </tr> <tr> <td>All</td> <td>0.25</td> <td>3.4</td>     <td>3.7</td> <td>28.7</td> </tr>  <tr> <td>Succeeded</td> <td>1.67</td> <td>3.4</td>     <td>3.7</td> <td>28.7</td> </tr> <tr> <td>Failed</td> <td>0.25</td> <td>1.64</td>     <td>1.7</td> <td>3.3</td> </tr></table>        |

**Comment by hcorson-dosch-usgs** [comment]
```r
> failed_runs <- p2_glm_uncalibrated_runs %>% filter(glm_success==FALSE) %>% group_by(site_id) %>% summarize(n_failed_runs = n())
> failed_runs %>% head(20)
# A tibble: 20 × 2
   site_id         n_failed_runs
   <chr>                   <int>
 1 nhdhr_114545137             2
 2 nhdhr_120018012             1
 3 nhdhr_120018078             1
 4 nhdhr_120018083             1
 5 nhdhr_120018485             1
 6 nhdhr_120018771             1
 7 nhdhr_120019286             1
 8 nhdhr_120020202             1
 9 nhdhr_120020261             1
10 nhdhr_120020265             1
11 nhdhr_120020305             1
12 nhdhr_120020396             1
13 nhdhr_120020455             1
14 nhdhr_120020615             1
15 nhdhr_120020620             1
16 nhdhr_120020682             1
17 nhdhr_120020716            12
18 nhdhr_120871578             2
19 nhdhr_121840587             1
20 nhdhr_132542729             1
> failed_runs %>% slice(21:40)
# A tibble: 20 × 2
   site_id                                    n_failed_runs
   <chr>                                              <int>
 1 nhdhr_132542847                                        1
 2 nhdhr_159101759                                       17
 3 nhdhr_34123198                                         1
 4 nhdhr_34123335                                         1
 5 nhdhr_34123830                                         1
 6 nhdhr_34126978                                         1
 7 nhdhr_34127736                                         2
 8 nhdhr_34128199                                         1
 9 nhdhr_34128217                                         1
10 nhdhr_34128269                                         1
11 nhdhr_34132253                                         1
12 nhdhr_52c8a174-e228-410d-8bf0-b60a24e1e08e             1
13 nhdhr_59745939                                        10
14 nhdhr_81003721                                         4
15 nhdhr_81003725                                         1
16 nhdhr_82957691                                         1
17 nhdhr_84624244                                         1
18 nhdhr_91574547                                        10
19 nhdhr_91596709                                         1
20 nhdhr_91681923                                         1
> failed_runs %>% slice(41:42)
# A tibble: 2 × 2
  site_id        n_failed_runs
  <chr>                  <int>
1 nhdhr_91684249            18
2 nhdhr_91692063             1
```

**Comment by hcorson-dosch-usgs** [comment]
Continuing to document this run.

If we group by site_id and filter to fully successful runs (n=18):
  * For 3541/3583 lakes, GLM model runs for each of the 3 time periods succeeded for all 6 GCMs. 
  * That's 63,738 runs (3541 * 3 time-period * 6 GCMs)

If we group by site_id and gcm and filter to fully successful runs (n=3) [as we do to package up the output]:
  * For 3579/3583 lakes, GLM model runs for each of the 3 time periods succeeded for at least 1/6 GCMs. 
  * That's 64,281 runs (3541 * 3 time-period * 1-6 GCMs)
  * That's 21,427 unique lake-gcm pairs for which there are formatted output files in `'caldera/projects/usgs/water/iidd/datasci/lake-temp/lake-temperature-process-models/3_extract/out'`
    * @lindsayplatt - Those ^ files are ready to be run through `lake-temperature-out`

For 42/3583 lakes, at least one GLM run for one GCM and one time period failed
  * The mean lake depth for those 42 lakes is 1.6 meters

For 33/3583 lakes, _only_ one GLM run for one GCM and one time period failed (1/18 runs failed)
  * The mean lake depth for those 33 lakes is 1.9 meters

For 9/3583 lakes, more than one GLM run for one GCM and one time period failed (> 1/18 runs failed)
  * The mean lake depth for those 9 lakes is 0.7 meters

For 1/3583 lakes, all 18 runs (3 time periods * 6 GCMs) failed
  * The lake depth for that lake is 0.3048 meters

**Reply by lindsayplatt** [reply]
Seeing the mean lake depth for these different categorizations of failures is definitely helpful :) Those shallow lakes are not playing nicely with our models!

**Comment by hcorson-dosch-usgs** [comment]
Some plots looking at the hypsography, GLM layers timeseries and GLM ice timeseries for failed runs. For those lakes w/ > 1 failed runs, I'm currently only plotting these results for the first of the failed runs. Since there are 42 lakes with failed runs, there are 42 plots, saved [here](https://doimspp.sharepoint.com/sites/IIDDStaff/Shared%20Documents/Forms/AllItems.aspx?RootFolder=%2Fsites%2FIIDDStaff%2FShared%20Documents%2FProject%20%2D%20Modeling%20Lakes%2FGLM%2Flake%2Dtemperature%2Dprocess%2Dmodels%2Fplots%5F220415%5Frun&FolderCTID=0x012000D4B31FA46E999B4A91C6B2D7E9D7A716)

The failed simulations represented in the plots (42/110 failed simulations) seem to fall into 3 categories:

1) Layers abruptly drop to 1 - (17/42 lakes)
    * Either right away (4/17 lakes) - 1 to 18 failed runs per lake
![image](https://user-images.githubusercontent.com/54007288/165183865-c540be2f-2513-476d-b809-aeb372c74cfb.png)
    * Or later (13/17 lakes) - 1 - 12 failed runs per lake
![image](https://user-images.githubusercontent.com/54007288/165182300-0b8f3d9b-a451-4fe0-ae76-141c58c2c54f.png)

2) Ice exceeds lake depth --> GLM drops to one layer. `nhdhr_34132253` - (22 lakes) - 21 w/ 1 failed run, 1 w/ 2 failed runs
![image](https://user-images.githubusercontent.com/54007288/165181188-1d77a3ec-bb64-4ebb-98d3-7225c11d09e8.png)
    * a more extreme example:
![image](https://user-images.githubusercontent.com/54007288/165181379-8ef28697-987e-4a52-9c07-992ff2c1a6ab.png)

3) Layers drop abruptly, then recover, but ice formation stops.  (3 lakes) - `nhdhr_91696709`, `nhdhrhr_82957691`, 'nhdhr_132432729 - all only 1 failed run
![image](https://user-images.githubusercontent.com/54007288/165181266-944225f0-dd18-43bc-99c3-cc8b783f63e8.png)

**Comment by hcorson-dosch-usgs** [comment]
Also made some plots for the 9 sites with >1 failed run. Plots [here](https://doimspp.sharepoint.com/:f:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Lakes/GLM/lake-temperature-process-models/plots_220415_run/[Multiple_failed_runs](https://doimspp.sharepoint.com/:f:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Lakes/GLM/lake-temperature-process-models/plots_220415_run/Multiple_failed_runs?csf=1&web=1&e=iwMoGY)?csf=1&web=1&e=iwMoGY). Plotting just the ice thickness and layers count:
![image](https://user-images.githubusercontent.com/54007288/165203579-de6d3731-510e-4678-aa42-d2404c2d2425.png)
![image](https://user-images.githubusercontent.com/54007288/165203628-779298ac-7ce7-4c92-b5f5-fe312b6ad25c.png)

**Comment by hcorson-dosch-usgs** [comment]
Update -- Yesterday I did two model builds on Tallgrass:
1) I re-ran all of the lake models for the 42 lakes that had failed model runs
2) I re-ran all 64,494 models

In both instances, the exact same 110 model runs failed. When I ran `tar_make()` on the `3_extract` targets after the full model run, `targets` skipped all of the targets, which indicates that the contents of [`p2_glm_uncalibrated_run_groups`](https://github.com/USGS-R/lake-temperature-process-models/blob/main/2_run.R#L44-L53) were unchanged and none of the output for the fully successful runs (successful = all 3 time periods ran for a single lake-gcm pair) had changed (as captured in the `export_fl_hash` field).

For a bit more context I checked where the 42 lakes with failed runs fall spatially -- They are in all 4 tiles, and spread across 25 cells.

Here's how those 110 failed runs break out by GCM:
![image](https://user-images.githubusercontent.com/54007288/167166621-0cfbdb2e-d6b4-44e9-8914-f50393d8c724.png)

And by time period:
![image](https://user-images.githubusercontent.com/54007288/167167123-a7f2188c-3936-4a4b-8c7f-3010f762e375.png)

**Comment by hcorson-dosch-usgs** [comment]
@lindsayplatt  Updated run results using [the updated nml](https://github.com/USGS-R/lake-temperature-model-prep/issues/339) and the [new grouping](https://github.com/USGS-R/lake-temperature-process-models/pull/47) to only export output for lakes for which all 18 runs (6 GCMs * 3 time periods) succeeded:

There are 3569 lakes. That's 64,242 runs. 64131/64242 runs succeeded. 111 failed. The failed runs were across 43 lakes. That means that we are exporting output for 3,526 lakes.

Here are details on lakes for which runs failed:
```r
> failed_runs %>% slice(1:20)
# A tibble: 20 × 2
   site_id         n_failed_runs
   <chr>                   <int>
 1 nhdhr_114545137             2
 2 nhdhr_120018012             1
 3 nhdhr_120018078             1
 4 nhdhr_120018083             1
 5 nhdhr_120018485             1
 6 nhdhr_120018615             1
 7 nhdhr_120018771             1
 8 nhdhr_120019286             1
 9 nhdhr_120020202             1
10 nhdhr_120020261             1
11 nhdhr_120020265             1
12 nhdhr_120020305             1
13 nhdhr_120020455             1
14 nhdhr_120020615             1
15 nhdhr_120020620             1
16 nhdhr_120020682             1
17 nhdhr_120020716            12
18 nhdhr_120871578             2
19 nhdhr_121840587             1
20 nhdhr_123112869             1
> failed_runs %>% slice(21:40)
# A tibble: 20 × 2
   site_id                                    n_failed_runs
   <chr>                                              <int>
 1 nhdhr_132542729                                        1
 2 nhdhr_132542847                                        1
 3 nhdhr_159101759                                       17
 4 nhdhr_32671566                                         1
 5 nhdhr_34123198                                         1
 6 nhdhr_34123335                                         1
 7 nhdhr_34123830                                         1
 8 nhdhr_34126978                                         1
 9 nhdhr_34127736                                         2
10 nhdhr_34128199                                         1
11 nhdhr_34128217                                         1
12 nhdhr_34128269                                         1
13 nhdhr_34132253                                         1
14 nhdhr_52c8a174-e228-410d-8bf0-b60a24e1e08e             1
15 nhdhr_59745939                                        10
16 nhdhr_81003721                                         4
17 nhdhr_81003725                                         1
18 nhdhr_82957691                                         1
19 nhdhr_84624244                                         1
20 nhdhr_91574547                                        10
> failed_runs %>% slice(41:43)
# A tibble: 3 × 2
  site_id        n_failed_runs
  <chr>                  <int>
1 nhdhr_91596709             1
2 nhdhr_91681923             1
3 nhdhr_91684249            18
> dput(failed_runs$site_id)
c("nhdhr_114545137", "nhdhr_120018012", "nhdhr_120018078", "nhdhr_120018083",
  "nhdhr_120018485", "nhdhr_120018615", "nhdhr_120018771", "nhdhr_120019286",
  "nhdhr_120020202", "nhdhr_120020261", "nhdhr_120020265", "nhdhr_120020305",
  "nhdhr_120020455", "nhdhr_120020615", "nhdhr_120020620", "nhdhr_120020682",
  "nhdhr_120020716", "nhdhr_120871578", "nhdhr_121840587", "nhdhr_123112869",
  "nhdhr_132542729", "nhdhr_132542847", "nhdhr_159101759", "nhdhr_32671566",
  "nhdhr_34123198", "nhdhr_34123335", "nhdhr_34123830", "nhdhr_34126978",
  "nhdhr_34127736", "nhdhr_34128199", "nhdhr_34128217", "nhdhr_34128269",
  "nhdhr_34132253", "nhdhr_52c8a174-e228-410d-8bf0-b60a24e1e08e",
  "nhdhr_59745939", "nhdhr_81003721", "nhdhr_81003725", "nhdhr_82957691",
  "nhdhr_84624244", "nhdhr_91574547", "nhdhr_91596709", "nhdhr_91681923",
  "nhdhr_91684249")
```

**Comment by hcorson-dosch-usgs** [comment]
@lindsayplatt - Here's the requested brief summary of the latest fresh GCM and NLDAS runs:

_________________________________________________________________

### GCM model runs

The results were identical to those shared in [the previous comment](https://github.com/USGS-R/lake-temperature-process-models/discussions/45#discussioncomment-2763238) - Out of 64,242 runs (3,569 lakes * 3 time periods * 6 GCMs), 64,131 succeeded (99.8%). 111 runs failed. The failed runs were across 43 lakes. That means that we exported output for 3,526 lakes. That's an overall success rate of 98.8%.

The mean lake depth across the 43 lakes is 1.7 meters. Minimum lake depth is 0.3048 m, maximum lake depth is 5.2 m.
_________________________________________________________________

### NLDAS model runs

The results were identical to those shared [in the PR](https://github.com/USGS-R/lake-temperature-process-models/pull/48#issuecomment-1128137907). Out of 3,569 runs (3,569 lakes), 3,558 succeeded (99.7%). Runs for 11 lakes failed. We exported output for 3,558 lakes.

The mean lake depth across the 11 lakes for which NLDAS runs failed is 0.8 meters. Minimum lake depth is 0.3048 m, maximum lake depth is 1.5 m.

**Reply by jordansread** [reply]
It may be useful to know what the overlap of successful lakes are between these two, since I'm assuming perhaps all of the failed NLDAS lakes also failed for at least one or more of the GCM sims, but perhaps that isn't the case. Since the dataset of lakes relevant to Gretchen's group (for example) is the subset of lakes that worked for all (NLDAS, all GCM sims). Is that number 3,526? or a smaller number?

**Reply by hcorson-dosch-usgs** [reply]
There were 4 lakes for which the NLDAS run failed but the GCM runs did not. 
There were 36 lakes for which the GCM runs failed but the NLDAS run did not.
There were 7 lakes for which both NLDAS and GCM runs failed.

That means there were either NLDAS or GCM failures across 47 lakes. So **we have NLDAS and all GCM simulation results for 3,522 lakes**. 3522/3569 = 98.7% of modeled lakes.

```R
> gcm_failed_sites <- failed_runs$site_id
> nldas_failed_sites <- p2_nldas_glm_uncalibrated_runs %>% filter(glm_success==FALSE) %>% pull(site_id)
> length(dput(nldas_failed_sites[!(nldas_failed_sites %in% gcm_failed_sites)]))
c("nhdhr_117648183", "nhdhr_120019262", "nhdhr_80997655", "nhdhr_82955915"
)
[1] 4
> length(dput(nldas_failed_sites[nldas_failed_sites %in% gcm_failed_sites]))
c("nhdhr_114545137", "nhdhr_120020716", "nhdhr_159101759", "nhdhr_59745939", 
"nhdhr_81003721", "nhdhr_81003725", "nhdhr_91684249")
[1] 7
> length(dput(gcm_failed_sites[!(gcm_failed_sites %in% nldas_failed_sites)]))
c("nhdhr_120018012", "nhdhr_120018078", "nhdhr_120018083", "nhdhr_120018485", 
"nhdhr_120018615", "nhdhr_120018771", "nhdhr_120019286", "nhdhr_120020202", 
"nhdhr_120020261", "nhdhr_120020265", "nhdhr_120020305", "nhdhr_120020455", 
"nhdhr_120020615", "nhdhr_120020620", "nhdhr_120020682", "nhdhr_120871578", 
"nhdhr_121840587", "nhdhr_123112869", "nhdhr_132542729", "nhdhr_132542847", 
"nhdhr_32671566", "nhdhr_34123198", "nhdhr_34123335", "nhdhr_34123830", 
"nhdhr_34126978", "nhdhr_34127736", "nhdhr_34128199", "nhdhr_34128217", 
"nhdhr_34128269", "nhdhr_34132253", "nhdhr_52c8a174-e228-410d-8bf0-b60a24e1e08e", 
"nhdhr_82957691", "nhdhr_84624244", "nhdhr_91574547", "nhdhr_91596709", 
"nhdhr_91681923")
[1] 36
> length(unique(c(nldas_failed_sites,gcm_failed_sites)))
[1] 47
> length(p1_site_ids) - length(unique(c(nldas_failed_sites,gcm_failed_sites)))
[1] 3522
````

**Comment by lindsayplatt** [comment]
Noting an issue with a MN lake that is causing my `lake-temperature-out` code to fail. Looks like this site ended up with only 1 layer for all of 1980-2000 in GFDL. I think maybe it should be flagged _before_ the export so that it isn't included? Skipping this lake in my code for now!

```r
# Using files from Caldera's lake-temperature-process-model dir
wtr_bad_contemporary <- arrow::read_feather('3_extract/out/GLM_nhdhr_114542607_GFDL.feather')
wtr_fine <- arrow::read_feather('3_extract/out/GLM_nhdhr_114542607_CNRM.feather')
```

**Reply by hcorson-dosch-usgs** [reply]
Strange -- I'm confused as to how that even got marked as a successful run. Here's the raw output, before it's filtered to the correct dates:
![image](https://user-images.githubusercontent.com/54007288/170314979-dfeaae2e-2658-43db-acf9-f616887377c2.png)

**Reply by lindsayplatt** [reply]
Weird that it looks OK but then suddenly on Oct 8th gets tons of ice. Seems like we may need to double check the checks :)

**Reply by hcorson-dosch-usgs** [reply]
Typically [this function](https://github.com/USGS-R/lake-temperature-process-models/blob/main/2_run/src/run_glm3.R#L9) fails when trying to extract the output and produce that ^ raw output file if there are NA values.

**Reply by lindsayplatt** [reply]
Do we need to add something like `filter(n_layers > 1)` to that to force a failure?

**Comment by hcorson-dosch-usgs** [comment]
@lindsayplatt - Here's the requested brief summary of the latest GCM run (conducted w/ batching of runs by lake), and a comparison w/ NLDAS results for Minnesota.

Interestingly, the batching (fewer, longer targets) has sped up the model run. With 4,584 lakes, the pipeline ran through completion of `p2_gcm` targets in just under 4 hours (though it could skip ~1000 branches that had been tested earlier), and through `p3_gcm` targets in 4.6 hours.
_________________________________________________________________

### GCM model runs

Out of 82,512 runs (4,584 lakes * 3 time periods * 6 GCMs), 81,784 succeeded (99%). 728 runs failed. The failed runs were across 222 lakes. That means that we exported GCM GLM output for 4,362 lakes. That's an overall success rate of 95%.

The mean lake depth across the 222 lakes is 1.47 meters. Median lake depth is 1 meter, minimum lake depth is 0 m (`nhdhr_45361479`), maximum lake depth is 10 m (`nhdhr_120018479`).

9/728 failed runs failed with `glm_code` = 1 and `max_output_date` < `param_sim_stop` (no extraction error). 1/728 failded with `glm_code` = 134. 718/728 failed runs failed with extraction errors:
```r
p2_gcm_glm_uncalibrated_runs %>% group_by(extraction_error) %>% summarize(n=n())
# A tibble: 4 × 2
  extraction_error                                                       n
  <chr>                                                              <int>
1 NA temperature values                                                341
2 Number of layers drops to 1                                          376
3 Number of layers drops to 1, Maximum hice value exceeds lake depth     1
4 NA                                                                 81794
```

Note that the 'NA temperature values' error is triggered by [the extraction of GLM temperatures](https://github.com/USGS-R/lake-temperature-process-models/blob/main/2_run/src/run_glm3.R#L17) and precedes [the generation of the other two extraction errors](https://github.com/USGS-R/lake-temperature-process-models/blob/main/2_run/src/run_glm3.R#L33-L48). So lakes for which temperature data extraction failed (`NA temperature values`) may have also have excessive ice generation or only 1 simulation layer.
_________________________________________________________________

### Comparison to latest NLDAS model runs

In the [latest NLDAS run for the full footprint](https://github.com/USGS-R/lake-temperature-process-models/discussions/57#discussioncomment-2867189), model runs for 4539/4584 Minnesota lakes succeeded. Model runs failed for 45/4584 Minnesota lakes.  

There were 6 Minnesota lakes for which the NLDAS run failed but the GCM runs did not.
There were 183 Minnesota lakes for which the GCM runs failed but the NLDAS run did not.
There were 39 Minnesota lakes for which both NLDAS and GCM runs failed.

That means there were either NLDAS or GCM failures across 228 Minnesota lakes. 

**So we have NLDAS and all GCM simulation results for 4,356 Minnesota lakes. 4356/4584 = 95% of modeled Minnesota lakes.**

**Reply by lindsayplatt** [reply]
Excellent! I think this point `the batching (fewer, longer targets) has sped up the model run` speaks to how `targets` does have overhead associated with branching and reading/writing a bunch of files, even if they are smaller!

**Reply by hcorson-dosch-usgs** [reply]
Yes, agreed!

The GCM evaluation plots are [HERE](https://doimspp.sharepoint.com/:f:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Lakes/GLM/lake-temperature-process-models/GCM_evaluation_plots/MN_20220603?csf=1&web=1&e=cRkOv6). There were 808 evaluation sites (sites with >= 10 min_obs_dates, successful model runs, and matched pred-obs).
